### DISCUSSION TITLE
How best to load queries from not officially supported parser?

### DISCUSSION THREAD

**Discussion by david-perez** [nvim-treesitter/nvim-treesitter#6175] [Q&A]
### Describe the highlighting problem

I wanted to try out [tree-sitter-prr](https://github.com/bcspragu/tree-sitter-prr), so I followed the ["Adding parsers" instructions](https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#adding-parsers) in this repo and added:

```lua
local parser_config = require("nvim-treesitter.parsers").get_parser_configs()
parser_config.prr = {
        install_info = {
                url = "https://github.com/bcspragu/tree-sitter-prr",
                branch = "main",
                files = { "src/parser.c" },
        },
}
```

Per the sentence in the instructions:

> If Neovim does not detect your language's filetype by default, you can use Neovim's vim.filetype.add() to add a custom detection rule.

I also added:

```lua
vim.filetype.add({
        extension = {
                prr = "prr",
        },
})
```

This revealed (`:checkhealth nvim-treesitter`, `:InspectTree`) the parser was being loaded, but no syntax highlighting was being applied to the file.

From the ["Adding queries"](https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#adding-queries) instructions and other users experiencing the same issue[^1][^2][^3], I guessed that what I was missing is loading the [`queries/highlights.scm`](https://github.com/bcspragu/tree-sitter-prr/blob/main/queries/highlights.scm) file in my `rtp`.

- Is there a way that `nvim-treesitter` can do this? I don't understand why for officially supported parsers, the queries in this repo (under `queries/{language}/*`) are automagically loaded, but for a manually added parser, they aren't. The structure for an officially supported parser (e.g. [`tree-sitter/tree-sitter-cpp`](https://github.com/tree-sitter/tree-sitter-cpp)) seems the same to me as the one I was trying to add (queries are under the `queries/` directory), so I don't understand why for a manually added parser they aren't automatically loaded too.
- If there isn't a way that `nvim-treesitter` can take care of this, what's the recommended way of adding the queries to the `rtp`? Can the documentation call this out more explicitly? I tried manually adding to my `rtp` the root folder of my parser, the `queries/` folder, and/or the `highlights.scm` file, but none worked. In the end I couldn't figure it out and simply cloned the repo and ran `cp tree-sitter-prr/queries/highlights.scm ~/.local/share/nvim/lazy/nvim-treesitter/queries/prr/`, but of course this is far from ideal.

[^1]: [Installed a local parser in nvim-treesitter, but no highlight](https://www.reddit.com/r/neovim/comments/zelzvx/installed_a_local_parser_in_nvimtreesitter_but_no/)
[^2]: [Syntax highlighting for custom tree-sitter parser not working](https://www.reddit.com/r/neovim/comments/14lij1z/syntax_highlighting_for_custom_treesitter_parser/)
[^3]: [Custom Grammar not recognized as Highlighting](https://www.reddit.com/r/neovim/comments/125utk4/comment/je9dap6/)

### Example snippet that causes the problem

N/A

### Tree-sitter parsing result

N/A

### Example screenshot

N/A

### Output of `:checkhealth nvim-treesitter`

```markdown
N/A
```

### Output of `nvim --version`

```text
NVIM v0.9.2
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Compilation: /usr/bin/cc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wvla -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -fno-common -Wno-unused-result -Wimplicit-fallthrough -fdiagnostics-color=always -fstack-protector-strong -DUNIT_TESTING -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include/luajit-2.1 -I/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/build/src/nvim/auto -I/tmp/x.sh.fdf/neovim-0.9.2/build/include -I/tmp/x.sh.fdf/neovim-0.9.2/build/cmake.config -I/tmp/x.sh.fdf/neovim-0.9.2/src -I/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include -I/tmp/x.sh.fdf/neovim-0.9.2/.deps/usr/include

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```

**Comment by clason** [comment]
It's a common misunderstanding that you need nvim-treesitter for anything but _installing_ (downloading, compiling) parsers. This is false. Neovim will _automatically_ load
* compiled parsers in any `parser` directory on `runtimepath` (e.g., `parser/prr.so`)
* queries in any `queries` directory on `runtimepath` (e.g., `queries/prr/highlights.scm`).

In fact, when developing parsers and queries, it's much easier to ignore nvim-treesitter and just add the checked out repo to `runtimepath` using a short [makefile](https://github.com/neovim/tree-sitter-vimdoc/blob/master/Makefile) that compiles the parser and copies it in `.parser`.

As to how, see `:h 'runtimepath'`. TL;DR: You should add the directory that contains `queries` and `parser`, then it should work.

**Reply by clason** [reply]
Also, tree-sitter syntax highlighting is opt-in, so you need to manually (or in a `FileType` autocommand) start it with `vim.treesitter.start()`.
